kimera_vio_ros/src/interfaces/imu_interface.cpp

```diff
@@ -0,0 +1,67 @@
+#include "kimera_vio_ros/interfaces/imu_interface.hpp"
+
+namespace kimera_vio_ros
+{
+namespace interfaces
+{
+
+ImuInterface::ImuInterface(
+  rclcpp::Node::SharedPtr & node)
+: BaseInterface(node),
+  last_imu_timestamp_(0)
+{
+  this->registerImuSingleCallback(
+    std::bind(
+      &VIO::Pipeline::fillSingleImuQueue,
+      pipeline_,
+      std::placeholders::_1));
+
+  this->registerImuMultiCallback(
+    std::bind(
+      &VIO::Pipeline::fillMultiImuQueue,
+      pipeline_,
+      std::placeholders::_1));
+
+  callback_group_imu_ = node->create_callback_group(
+    rclcpp::callback_group::CallbackGroupType::MutuallyExclusive);
+  auto imu_opt = rclcpp::SubscriptionOptions();
+  imu_opt.callback_group = callback_group_imu_;
+
+  std::string imu_topic = "imu";
+  auto qos = rclcpp::SensorDataQoS();
+  imu_sub_ = node->create_subscription<Imu>(
+    imu_topic,
+    qos,
+    std::bind(
+      &ImuInterface::imu_cb,
+      this,
+      std::placeholders::_1),
+    imu_opt);
+}
+
+ImuInterface::~ImuInterface()
+{
+}
+
+void ImuInterface::imu_cb(const Imu::SharedPtr imu_msg)
+{
+  rclcpp::Time stamp(imu_msg->header.stamp);
+  if (stamp.nanoseconds() > last_imu_timestamp_.nanoseconds()) {
+    VIO::Timestamp timestamp = stamp.nanoseconds();
+    VIO::ImuAccGyr imu_accgyr;
+
+    imu_accgyr(0) = imu_msg->linear_acceleration.x;
+    imu_accgyr(1) = imu_msg->linear_acceleration.y;
+    imu_accgyr(2) = imu_msg->linear_acceleration.z;
+    imu_accgyr(3) = imu_msg->angular_velocity.x;
+    imu_accgyr(4) = imu_msg->angular_velocity.y;
+    imu_accgyr(5) = imu_msg->angular_velocity.z;
+
+    this->imu_single_callback_(VIO::ImuMeasurement(timestamp, imu_accgyr));
```

Comment:
@ToniRV , I'm still having some issues with `No IMU data from last frame timestamp:` when using `parallel_run=true`. Any suggestions other than setting parallel_run to false. I'd like to keep `pipline_.spin()` in it's own async thread.

```
$ ros2 launch kimera_vio_ros kimera_ros_realsense.launch.xml 
[INFO] [launch]: All log files can be found below /home/ruffsl/.ros/log/2020-03-02-18-45-16-443174-dox-12704
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [kimera_vio_ros-1]: process started with pid [12715]
[kimera_vio_ros-1] I0302 18:45:16.720127 12715 CameraParams.cpp:31] Parsing camera parameters for: left_cam
[kimera_vio_ros-1] I0302 18:45:16.787691 12715 CameraParams.cpp:31] Parsing camera parameters for: right_cam
[kimera_vio_ros-1] I0302 18:45:16.788328 12715 DataProviderInterface.cpp:96] Running dataset between frame 50 and frame 10000
[kimera_vio_ros-1] I0302 18:45:16.796392 12715 RegularVioBackEnd.cpp:113] Using Regular VIO backend.
[kimera_vio_ros-1] I0302 18:45:17.805269 12734 PipelineModule.h:163] Module: Data Provider - Spinning.
[kimera_vio_ros-1] W0302 18:45:17.817641 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
[kimera_vio_ros-1] W0302 18:45:17.818171 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
[kimera_vio_ros-1] W0302 18:45:17.818605 12734 DataProviderModule.cpp:97] No IMU data from last frame timestamp: 1583203517581242112 to timestamp: 1583203517614594304
...
```

Reply:
I think I was encountering an error with an uninitialized  '_P' camera param given my setting of `rectified=true` (which should be true given that's what the realsense publishes), but found out I was using an incorrect nominalBaseline before that was failing the rectification setup that would normally initialized  '_P'. However, now I'm getting a rather strange issue with the pipline setup with Eigen:

```
$ ros2 launch kimera_vio_ros kimera_ros_realsense.launch.xml 
[INFO] [launch]: All log files can be found below /home/ruffsl/.ros/log/2020-03-02-21-18-24-670948-dox-3734
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [kimera_vio_ros-1]: process started with pid [3745]
[kimera_vio_ros-1] I0302 21:18:24.941282  3745 CameraParams.cpp:31] Parsing camera parameters for: left_cam
[kimera_vio_ros-1] I0302 21:18:24.993412  3745 CameraParams.cpp:31] Parsing camera parameters for: right_cam
[kimera_vio_ros-1] I0302 21:18:24.994200  3745 DataProviderInterface.cpp:96] Running dataset between frame 50 and frame 10000
[kimera_vio_ros-1] I0302 21:18:25.002403  3745 RegularVioBackEnd.cpp:113] Using Regular VIO backend.
[kimera_vio_ros-1] I0302 21:18:25.010413  3759 PipelineModule.h:163] Module: Data Provider - Spinning.
[kimera_vio_ros-1] I0302 21:18:25.198184  3759 Pipeline.cpp:780] Frontend launched (parallel_run set to 1).
[kimera_vio_ros-1] I0302 21:18:25.198218  3775 PipelineModule.h:163] Module: VioFrontEnd - Spinning.
[kimera_vio_ros-1] I0302 21:18:25.198231  3759 Pipeline.cpp:511] ------------------- Initialize Pipeline with frame k = 1--------------------
[kimera_vio_ros-1] stereo_vio_node: /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:374: void Eigen::PlainObjectBase<Derived>::resizeLike(const Eigen::EigenBase<OtherDerived>&) [with OtherDerived = Eigen::Matrix<double, 6, -1>; Derived = Eigen::Matrix<double, 6, 1>]: Assertion `other.rows() == 1 || other.cols() == 1' failed.
```

Reply:
Resolved with https://github.com/MIT-SPARK/Kimera-VIO/issues/37